sycl-fusion/passes/internalization/Internalization.cpp

```diff
@@ -0,0 +1,670 @@
+  auto *ArrTy = ArrayType::get(Ty, LocalSize);
+  auto *Alloca = Builder.CreateAlloca(ArrTy, PtrTy->getAddressSpace());
+  auto *Ptr = Builder.CreateInBoundsGEP(
+      ArrTy, Alloca, {Builder.getInt64(0), Builder.getInt64(0)});
+  Arg->replaceAllUsesWith(Ptr);
+  Alloca->mutateType(PointerType::getWithSamePointeeType(
+      cast<PointerType>(Alloca->getType()), AS));
+  Ptr->mutateType(PointerType::getWithSamePointeeType(
+      cast<PointerType>(Ptr->getType()), AS));
+  return Ptr;
+}
+
+Function *SYCLInternalizerImpl::promoteFunctionArgs(
+    Function *OldF, ArrayRef<std::size_t> PromoteToLocal, bool CreateAllocas,
+    bool KeepOriginal) const {
+  constexpr unsigned AddressSpaceBitWidth{32};
+
+  auto *NewAddrspace = ConstantAsMetadata::get(ConstantInt::get(
+      IntegerType::get(OldF->getContext(), AddressSpaceBitWidth), AS));
+
+  // We first declare the promoted function with the new signature.
+  llvm::Function *NewF =
+      getPromotedFunctionDeclaration(OldF, PromoteToLocal, AS,
+                                     /*ChangeTypes*/ !CreateAllocas);
+
+  // Clone the old function into the new function.
+  {
+    ValueMap<const Value *, WeakTrackingVH> VMap;
+    for (auto SrcI = OldF->arg_begin(), DestI = NewF->arg_begin();
+         SrcI != OldF->arg_end(); ++SrcI, ++DestI) {
+      DestI->setName(SrcI->getName()); // Copy argument names.
+      VMap[&*SrcI] = &*DestI;          // Map homologous arguments.
+    }
+    // We can omit the vector of returns.
+    SmallVector<ReturnInst *> Returns;
+    CloneFunctionInto(NewF, OldF, VMap, CloneFunctionChangeType::GlobalChanges,
+                      Returns);
+  }
+
+  // Update each of the values to promote.
+  for (auto I : enumerate(PromoteToLocal)) {
+    const auto LocalSize = I.value();
+    if (LocalSize == 0) {
+      continue;
+    }
+    const auto Index = I.index();
+    Value *Arg{NewF->getArg(Index)};
+    if (!isa<PointerType>(Arg->getType()) ||
+        cast<Argument>(Arg)->hasByValAttr()) {
+      // Omit non-pointer and byval arguments.
+      continue;
+    }
+    if (CreateAllocas) {
+      Arg = replaceByNewAlloca(cast<Argument>(Arg), AS, LocalSize);
+    }
+    promoteValue(Arg, LocalSize);
+  }
+
+  {
+    constexpr StringLiteral KernelArgAddrSpaceMD{"kernel_arg_addr_space"};
+    if (auto *AddrspaceMD =
+            dyn_cast_or_null<MDNode>(NewF->getMetadata(KernelArgAddrSpaceMD))) {
+      // If we have kernel_arg_addr_space metadata in the original function,
+      // we should update it in the new one.
+      SmallVector<Metadata *> NewInfo{AddrspaceMD->op_begin(),
+                                      AddrspaceMD->op_end()};
+      for (auto I : enumerate(PromoteToLocal)) {
+        if (I.value() == 0) {
+          continue;
+        }
+        const auto Index = I.index();
+        if (const auto *PtrTy =
+                dyn_cast<PointerType>(NewF->getArg(Index)->getType())) {
+          if (PtrTy->getAddressSpace() == 3) {
```

Comment:
Can we at least have a comment about what this magic number `3` means? Ideally, we should get rid of it